### PR TITLE
Add documentation for dot product attention

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

Add documentation for dot product attention.

## Type of change

- [x] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

- Add documentation, example scripts, benchmarks for dot product attention. Discuss the details of supported backends, their differences, and the nuances of other features such as QKV layouts, mask types, and FP8 attention.

# Checklist:

- [x ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x ] The functionality is complete
- [x ] I have commented my code, particularly in hard-to-understand areas
- [x ] I have made corresponding changes to the documentation
- [x ] My changes generate no new warnings
- [x ] I have added tests that prove my fix is effective or that my feature works
- [x ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: benchmarks/attention/benchmark_attention.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    cudnn_times = []
+    flash_times = []
+    warmup_iters = 3
+    for i in range(num_iters+warmup_iters):
+        if i == warmup_iters:
+            torch.cuda.cudart().cudaProfilerStart()
+
+        if fused_attn_supported:
+            torch.cuda.synchronize()
+            fused_attn_start = time.time()
```
Comment (zlsh80826) [inline benchmarks/attention/benchmark_attention.py:62]: `time.perf_counter` is preferred and more accurate for measuring the duration then `time.time`.

File: benchmarks/attention/benchmark_attention.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        prof_cmd = f"""nsys profile \
+                --capture-range=cudaProfilerApi \
+                --capture-range-end=stop-shutdown \
+                --force-overwrite=true \
+                --output=prof_{model} \
+                python -c "import benchmark_attention; \
+                benchmark_attention.benchmark_dot_product_attention(\
+                '{model}', {fused_attn_supported}, {flash_attn_supported})" """
+        os.system(prof_cmd)
```
Comment (zlsh80826) [inline benchmarks/attention/benchmark_attention.py:159]: `os.system` is not safe. `subprocess` module is preferred than `os.system`. See https://docs.python.org/3/library/os.html#os.system

File: benchmarks/attention/benchmark_attention.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    flash_times = []
+    warmup_iters = 3
+    for i in range(num_iters+warmup_iters):
+        if i == warmup_iters:
+            torch.cuda.cudart().cudaProfilerStart()
+
+        if fused_attn_supported:
+            torch.cuda.synchronize()
+            fused_attn_start = time.time()
+            fused_attn_fwd, fused_attn_bwd = _run_dot_product_attention(
```
Comment (zlsh80826) [inline benchmarks/attention/benchmark_attention.py:63]: We usually like to run several times then add the synchronize at the end and computes the average when benchmarking. This can overlap the CPU launch time and GPU compute to reflect the real workload. Adding synchronize per iter will measure the CPU launch/pre-processing time as those time is usually overlapped in the real workload.

**Review by zlsh80826 — COMMENTED** [review]
Thanks for making the attention document, it's very well described. I'm still reviewing the jupter notebook will try to finish it this week. I've left a few comments regarding the benchmark scripts.

File: docs/examples/attention/attention.ipynb [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+     "output_type": "stream",
+     "text": [
+      "Running test_0...\n",
+      "Running test_1...\n",
+      "Running test_2...\n",
+      "Running test_3...\n",
+      "        cuDNN fwd+bwd (ms)  flash-attn fwd+bwd (ms)  cuDNN vs flash speedup\n",
+      "test_0              0.0644                   0.0846                  1.3145\n",
+      "test_1              0.5453                   0.7626                  1.3986\n",
+      "test_2              1.2363                   0.0000                  0.0000\n",
```
Comment (zlsh80826) [inline docs/examples/attention/attention.ipynb:130]: Is the runtime = 0ms here expected?

File: docs/examples/attention/attention.ipynb [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "id": "db92997e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```
Comment (zlsh80826) [inline docs/examples/attention/attention.ipynb:122]: Showing GPU arch here would be better for users to reproduce

**Review by zlsh80826 — APPROVED** [review]
LGTM. The document is very well, list a few different points in terms of JAX.
1. JAX havn't support `NVTE_DEBUG` and `NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT`, ALIBI mask, FP8 attn.
2. `attention_mask` in JAX is only a tensor which encodes both `q` and `kv`, expected to have shape `[b, 1 | h, s_q, s_kv]`.

File: docs/examples/attention/attention.ipynb (hunk 4 above)
Comment (cyanguwa) [reply docs/examples/attention/attention.ipynb:130]: Runtime=0ms means that that particular config is not supported - in this case, by flash-attention. This is explained in the text above this code cell.

File: docs/examples/attention/attention.ipynb (hunk 5 above)
Comment (cyanguwa) [reply docs/examples/attention/attention.ipynb:122]: Added some prints for GPU info.
